spec/Candid.md

```diff
@@ -882,201 +882,150 @@ service { <name> : <functype>; <methtype>;* } <: service { <name> : <functype'>;
 
 ```
-<v> ~> <v'> : <t>
-----------------------------------------------------------------------------------
-variant { <nat> = <v> } ~> variant { <nat> = <v'> } : variant { <nat> = <t>; _;* }
+c[variant { <nat> = <t>;* _;* } <: variant { <nat> = <t'>;* _;* }](variant { <natN> = <v> })
+    = variant { <natN> = c[<tN> <: <t'N>](<v>) }
 ```
 
 
 #### References
 
-Function and services references coerce unconditionally (note that coercion is only applied when the input value has a subtype of the expected type):
-
-```
-------------------------------------------------------
-func <text>.<id> ~> func <text>.<id> : func <functype>
-```
-
-```
-------------------------------------------------------
-service <text> ~> service <text> : service <actortype>
-```
+Function and services reference values are untyped, so the coercion function is the identity here:
 
 ```
-------------------------------------------------
-principal <text> ~> principal <text> : principal
+c[func <functype> <: func <functype'>](func <text>.<id>) = func <text>.<id>
+c[service <actortype> <: service <actortype'>](service <text>) = service <text>
+c[principal <: principal](principal <text>) = principal <text>
 ```
 
 #### Tuple types
 
 Whole argument and result sequences are coerced with the same rules as tuple-like records. In particular, extra arguments are ignored, and optional parameters read as as `null` if the argument is missing or fails to coerce:
 
 ```
-record {<v>;*} ~> record {<v'>,*} : record {<t>;*}
---------------------------------------------------
-(<v>,*) ~> (<v'>,*) : (<t>,*)
+c[(<t>,*) <: (<t'>,*)]((<v>,*)) = (<v'>,*)
+    if c[record {<t>;*} ~> record {<t'>,*}](record {<v>;*}) = record {<v'>;*}
 ```
 
 
 ## Properties
 
-The relations above have certain properties. To express them, we need the relation `V : T`, expressing that `V` has inherently type `T`. Instead of defining this relation on its own, we take the roundtripping property below as its definition:
+The relations above have certain properties. As in the previous section, `<v> : <t>` means that <v> has inherently type `<t>`.
 
-* Correctness and completeness of decoding:
+* Reflexivity of subtyping:
   ```
-  (v : T) ⟺ (∃ v'. v' ~> v : T)
+  <t> <: <t>
   ```
 
-* Roundtripping:
+* Transitivity of subtyping:
   ```
-  (v : T) ⟺ v ~> v : T
+  <t> <: <t'>, <t'> <: <t''> ⇒ <t> <: <t'>
   ```
 
-* Uniqueness of decoding:
+* Roundtripping:
   ```
-  v ~> v1 : T ∧ v ~> v2 : T ⇒ v1 = v2
+  (<v> : <t>) ⟺ c[<t> <: <t>](<v>) = <v>
   ```
 
-* Soundness of subtyping:
+* Soundness of subtyping (or, alternatively, well-definedness of coercion)
   ```
-  T <: T' ⇒ ∀ v : T. ∃ v'. v ~> v' : T
+  <t> <: <t'> ⇒ ∀ <v> : <t>. v[<t> <: <t'>](<v>) : <t'>
   ```
 
 * Higher-order soundness of subtyping
   See <./IDL-Soundness.md>, with the following instantiations:
   ```
   s1 ~> s2 ≔ s2 <: s1
-  t1 <: t2 ≔ (∀ v. v : t1 ⇒ v ~> _ : t2 )
+  t1 <: t2 ≔ t1 <: t2
```

Comment:
I suppose you could just drop this line?

Reply:
Don’t think so, the left hand side is “parameters of IDL-Soundness”, the right hand side is “stuff defined here”. The punnig is mostly accidential.